src/main/java/org/spongepowered/api/util/command/spec/CommandSpec.java

```diff
@@ -0,0 +1,392 @@
+        return complete(source, args, new CommandContext());
+    }
+
+    @Override
+    public boolean testPermission(CommandSource source) {
+        return this.permission == null || source.hasPermission(this.permission);
+    }
+
+    /**
+     * Get a short, one-line description used with this command if any is present
+     *
+     * @return the short description.
+     */
+    @Override
+    public Optional<Text> getShortDescription(CommandSource source) {
+        return Optional.fromNullable(this.description);
+    }
+
+    /**
+     * Get the usage for this command appropriate for the provided command source.
+     *
+     * @param source The source
+     * @return the usage for the source
+     */
+    @Override
+    public Text getUsage(CommandSource source) {
+        Preconditions.checkNotNull(source, "source");
+        return this.args.getUsage(source);
+    }
+
+    /**
+     * Return a longer description for this command. This description is composed of at least all present of the short description, the usage
+     * statement, and the extended description
+     * @param source The source to get the extended description for
+     * @return the extended description
+     */
+    @Override
+    public Optional<Text> getHelp(CommandSource source) {
+        Preconditions.checkNotNull(source, "source");
+        TextBuilder builder = Texts.builder();
+        Optional<Text> desc = getShortDescription(source);
+        if (desc.isPresent()) {
+            builder.append(desc.get(), CommandMessageFormatting.NEWLINE_TEXT);
+        }
+        builder.append(getUsage(source));
+        if (this.extendedDescription != null) {
+            builder.append(CommandMessageFormatting.NEWLINE_TEXT, this.extendedDescription);
+        }
+        return Optional.of(builder.build());
+    }
+
+    @Override
+    public boolean equals(Object o) {
+        if (this == o) {
+            return true;
+        }
+        if (!(o instanceof CommandSpec)) {
+            return false;
+        }
+
+        CommandSpec that = (CommandSpec) o;
+
+        if (!this.args.equals(that.args)) {
+            return false;
+        }
+        if (!this.argumentParser.equals(that.argumentParser)) {
+            return false;
+        }
+        if (this.description != null ? !this.description.equals(that.description) : that.description != null) {
+            return false;
+        }
+        if (!this.executor.equals(that.executor)) {
+            return false;
+        }
+        if (this.extendedDescription != null ? !this.extendedDescription.equals(that.extendedDescription) : that.extendedDescription != null) {
+            return false;
+        }
+        if (this.permission != null ? !this.permission.equals(that.permission) : that.permission != null) {
+            return false;
+        }
+
+        return true;
+    }
+
+    @Override
+    public int hashCode() {
+        int result = this.args.hashCode();
```

Comment:
You can use `Objects.hashCode(Object...)` from Guava here to simplify this (we use this in a lot of classes already)